narciuch/assignment1/src/SinglyLinkedList.java

```diff
@@ -0,0 +1,84 @@
+public class SinglyLinkedList<T> {
+  private int size = 0;
+  private Node head;
+
+  public int getSize() {
+    return size;
+  }
+
+  /**
+   * Get element at specified index from the singly linked list.
+   * Indexing from 0 (i.e. head is element zero).
+   *
+   * @param elemIndex : index of desired element
+   * @return value of the the kth element in the list
+   */
+  public T getElemAtIndex(int elemIndex) {
+    if (elemIndex > size) {
+        return null;
+    }
+    Node currentNode = head;
+    for (int i = 0; i < elemIndex && currentNode.getNext() != null; i++) {
+      currentNode = currentNode.getNext();
+    }
+    return currentNode.getValue();
+  }
+
+  /**
+   * Finds the kth to last element of the list
+   *
+   * @param nrFromEnd: desired index from end of list.
+   *                   nrFromEnd of 0 corresponds to the last element in the list,
+   *                   nrFromEnd of 1 corresponds to the second to last element etc.
+   * @return value of the nrFromEnd element
+   */
+  public T getKthElemFromTail(int nrFromEnd) {
+    return getElemAtIndex(size - nrFromEnd - 1);
```

Comment:
Clever use of size to simplify code! What would the code look like if size was not available?

Reply:
Would be possible to traverse the list once every time this code is called, but the extra time to do this does not outweigh the small amount of space required to simply keep track of the size. 

As you mentioned in the group session, another possibility would have been to use a moving window of size k with 2 pointers.